### PR TITLE
Change AAE hashtree to buffer and batch write to LevelDB

### PR DESCRIPTION
**Pull request by jtuple**
Change AAE to batch write to LevelDB rather than write hash updates one at a time. While AAE writes are small and asynchronous, having 2x the writes (normal Riak write + AAE write) going into LevelDB has shown to have an impact on Riak throughput. This pull-request has hashtrees buffer writes in memory until there are 200 hash updates, and then send all writes to LevelDB as a single batch write. This has been benchmarked to show a noticeable performance improvement.

In addition to batch writing once every 200 writes, the hashtree will also always flush writes before updating a tree and performing a tree exchange. Thus, AAE semantics are identical between non-batched and batched approach.

If a `riak_kv_index_hashtree` crashes for some reason, and therefore crashes the `hashtree` processes, we may lose the last few writes buffered in RAM. This isn't a problem, as AAE is designed to be lossy anyway, and exchanges with other replicas and/or eventual re-building of the tree will sort things out.

This pull-request also makes the LevelDB options used by AAE configurable, as well as bumps the default write-buffer size from 1 MB to 4 MB.

### CODE REVIEW

File: src/hashtree.erl [hunk 1]
```diff
@@ -253,14 +257,33 @@ insert(Key, ObjHash, State, Opts) ->
...
+            State3 = maybe_flush_buffer(State2),
+            %% Dirty = gb_sets:add_element(Segment, State3#state.dirty_segments),
+            Dirty = bitarray_set(Segment, State3#state.dirty_segments),
+            State3#state{dirty_segments=Dirty};
         false ->
             State
     end.
 
+maybe_flush_buffer(State=#state{write_buffer_count=WCount}) ->
+    Threshold = 200,
```
Comment (joedevivo) [inline src/hashtree.erl:277]: Is there a reason we'd ever want this to be configurable?

**Comment by joedevivo** [comment]
update the @docs in hashtree.erl around line 82: "updated in real-time as writes occur"

File: src/hashtree.erl [hunk 2]
```diff
@@ -290,14 +313,16 @@ should_insert(HKey, Opts, State) ->
...
+    SnapState = snapshot(State2),
+    State3 = SnapState#state{dirty_segments=bitarray_new(NumSegments)},
+    {SnapState, State3}.
 
 -spec update_tree(hashtree()) -> hashtree().
 update_tree(State) ->
-    State2 = snapshot(State),
-    update_perform(State2).
+    State2 = flush_buffer(State),
+    State3 = snapshot(State2),
```
Comment (joedevivo) [inline src/hashtree.erl:329]: What happens if an insert comes in between flush_buffer/1 and snapshot/1?

Comment (jtuple) [inline src/hashtree.erl:329]: There is no "in between". `insert` and `update_tree` are both top-level functions that operate on a hash tree. You put a hash tree in, get a hash tree out. You have two options:

``` erlang
B = hashtree:insert(Key, Hash, A),
C = hashtree:update_tree(B).
```

or 

``` erlang
B = hashtree:update_tree(A),
C = hashtree:inesrt(Key, Hash, B)
```

The behavior is clear in either case.

File: src/hashtree.erl (hunk 1 above)
Comment (jtuple) [inline src/hashtree.erl:277]: I keep going back and forth on this one. My current opinion is hardcode it and QA things against a static value. How much you want to buffer or not has little to do with user workload, cluster size, etc. It's almost entirely just a "magic number" that we want to optimize against Riak / LevelDB.

**Comment by joedevivo** [comment]
+1 to merge, once that documentation is fixed.

**Pull request merged**